README.rst
***********************
eagerx_examples package
***********************

.. image:: https://img.shields.io/badge/License-Apache_2.0-blue.svg
   :target: https://opensource.org/licenses/Apache-2.0
   :alt: license

.. image:: https://img.shields.io/badge/code%20style-black-000000.svg
   :target: https://github.com/psf/black
   :alt: codestyle

.. image:: https://github.com/eager-dev/eagerx_examples/actions/workflows/ci.yml/badge.svg?branch=master
  :target: https://github.com/eager-dev/eagerx_examples/actions/workflows/ci.yml
  :alt: Continuous Integration

.. contents:: Table of Contents
    :depth: 2

What is the *eagerx_examples* package?
======================================
This repository/package contains coding examples for EAGERx.
EAGERx (Engine Agnostic Gym Environments for Robotics) enables users to easily define new tasks, switch from one sensor to another, and switch from simulation to reality with a single line of code by being invariant to the physics engine.

`The core repository is available here <https://github.com/eager-dev/eagerx>`_.

`Full documentation and tutorials (including package creation and contributing) are available here <https://eagerx.readthedocs.io/en/master/>`_.

Installation
============

You can install the package using pip:

.. code:: shell

    pip3 install eagerx-examples

.. note::
    EAGERx depends on a minimal ROS installation. Fortunately, you **can** use eagerx anywhere as you would any python package,
    so it does **not** impose a ROS package structure on your project.

Cite EAGERx
===========
If you are using EAGERx for your scientific publications, please cite:

.. code:: bibtex

    @article{eagerx,
        author  = {van der Heijden, Bas and Luijkx, Jelle, and Ferranti, Laura and Kober, Jens and Babuska, Robert},
        title = {EAGER: Engine Agnostic Gym Environment for Robotics},
        year = {2022},
        publisher = {GitHub},
        journal = {GitHub repository},
        howpublished = {\url{https://github.com/eager-dev/eagerx}}
    }

Acknowledgements
=================
EAGERx is funded by the `OpenDR <https://opendr.eu/>`_ Horizon 2020 project.


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.1.5 (2022-03-28)
### Fix
* Eagerx api changes ([`85edc8e`](https://github.com/eager-dev/eagerx_examples/commit/85edc8ea3c39a370ac1e7654c5cf651f7fcd9520))
* Api changes eagerx ([`726527c`](https://github.com/eager-dev/eagerx_examples/commit/726527cbfc9a177a1c552fcaa9287c7281783e05))


tests/test_eagerx_examples.py
def test_example_full_state(name="test_full_state", eps=1, eval_eps=1, eps_length=1):
    from eagerx_examples.example_full_state import example_full_state

    example_full_state(name=name, eps=eps, eval_eps=eval_eps, eps_length=eps_length)


def test_example_pid(name="test_pid_only", eps=1, eval_eps=1, eps_length=1):
    from eagerx_examples.example_pid_only import example_pid_only

    example_pid_only(name=name, eps=eps, eval_eps=eval_eps, eps_length=eps_length)


def test_example_classifier(name="test_classifier", eps=1, eval_eps=1, eps_length=1, classifier_epochs=1):
    from eagerx_examples.example_classifier import example_classifier

    example_classifier(name=name, eps=eps, eval_eps=eval_eps, eps_length=eps_length, classifier_epochs=classifier_epochs)


eagerx_examples/pid.py
from typing import Optional, List
from collections import deque
import numpy as np

# IMPORT ROS
from std_msgs.msg import Float32MultiArray

# IMPORT EAGERX
import eagerx.core.register as register
from eagerx.utils.utils import Msg
from eagerx.core.entities import Node, Processor, SpaceConverter
from eagerx.core.constants import process


class PidController(Node):
    @staticmethod
    @register.spec("PidController", Node)
    def spec(
        spec,
        name: str,
        rate: float,
        gains: List[float],
        y_range: Optional[List[float]] = None,
        u_range: Optional[List[float]] = None,
        index: int = 0,
        u0: float = 0.0,
        process: Optional[int] = process.NEW_PROCESS,
        color: Optional[str] = "grey",
    ):
        """
        PID controller.

        :param spec: Not provided by user.
        :param name: Node name
        :param rate: Rate at which callback is called.
        :param y_range: reference range [min, max]. This is not enforced, but merely used for space_converters.
        :param u_range: control range [min, max]. This is not enforced, but merely used for space_converters.
        :param gains: [Kp, Kd, Ki]
        :param index: Index (related to Float32MultiArray.data[index])
        :param u0: initial action
        :param process: {0: NEW_PROCESS, 1: ENVIRONMENT, 2: BRIDGE, 3: EXTERNAL}
        :param color: console color of logged messages. {'black', 'red', 'green', 'yellow', 'blue', 'magenta', 'cyan', 'white', 'grey'}
        :return:
        """
        # Performs all the steps to fill-in the params with registered info about all functions.
        spec.initialize(PidController)

        # Set default
        y_range = y_range if y_range else [-1.0, 1.0]
        u_range = u_range if u_range else [-1.0, 1.0]

        # Modify default node params
        params = dict(name=name, rate=rate, process=process, color=color, inputs=["y", "yref"], outputs=["u"])
        spec.config.update(params)

        # Modify custom node params
        spec.config.gains = gains
        spec.config.u0 = u0

        # Add converter & space_converter to outputs
        c = Processor.make("GetIndex_Float32MultiArray", index=index)
        spec.outputs.u.converter = c
        spec.outputs.u.space_converter = SpaceConverter.make(
            "Space_Float32MultiArray", [u_range[0]], [u_range[1]], dtype="float32"
        )

        # Add space_converter to inputs
        sc = SpaceConverter.make("Space_Float32MultiArray", [y_range[0]], [y_range[1]], dtype="float32")
        spec.inputs.y.space_converter = sc
        spec.inputs.yref.space_converter = sc
        spec.inputs.y.converter = c
        spec.inputs.yref.converter = c

    def initialize(self, gains, u0):
        self.controller = PID(u0=u0, kp=gains[0], kd=gains[1], ki=gains[2], dt=1 / self.rate)

    @register.states()
    def reset(self):
        self.controller.reset()

    @register.inputs(y=Float32MultiArray, yref=Float32MultiArray)
    @register.outputs(u=Float32MultiArray)
    def callback(self, t_n: float, y: Optional[Msg] = None, yref: Optional[Msg] = None):
        trig = y.msgs[-1].data[:2]
        y = np.arctan2(trig[1], trig[0])
        yref = yref.msgs[-1].data[0]
        u = self.controller.next_action(y, ref=yref)
        return dict(u=Float32MultiArray(data=np.array([u])))


class PID:
    def __init__(self, u0: float, kp: float, kd: float, ki: float, dt: float):
        self.u0 = u0
        self.kp = kp
        self.kd = kd
        self.ki = ki
        self.dt = dt

        self.F = [kp + ki * dt + kd / dt, -kp - 2 * kd / dt, kd / dt]

        self.window = None
        self.u = None

    def reset(self):
        self.u = self.u0
        self.window = deque(maxlen=3)

    def next_action(self, y: float, ref: float = 0.0):
        # Add error
        self.window.appendleft(ref - y)

        # Calculate action
        for idx, e in enumerate(self.window):
            self.u += self.F[idx] * e

        return self.u


eagerx_examples/converters.py
# ROS IMPORTS
from std_msgs.msg import Float32MultiArray

# RX IMPORTS
import eagerx.core.register as register
from eagerx import Processor
import numpy as np


class AngleDecomposition(Processor):
    MSG_TYPE = Float32MultiArray

    @staticmethod
    @register.spec("AngleDecomposition", Processor)
    def spec(spec, angle_idx: int = 0):
        spec.initialize(AngleDecomposition)
        spec.config.angle_idx = angle_idx

    def initialize(self, angle_idx=0):
        assert angle_idx == 0, "Converter currently only works when angle_idx = 0 and "
        self.angle_idx = angle_idx

    def convert(self, msg):
        if msg.data == []:
            return msg
        angle = msg.data[self.angle_idx]
        new_data = np.concatenate(([np.sin(angle), np.cos(angle)], msg.data[self.angle_idx + 1 :]), axis=0)
        return Float32MultiArray(data=new_data)


eagerx_examples/example_pid_only.py
def example_pid_only(name, eps, eval_eps, eps_length=200):
    # EAGERx imports
    from eagerx import Object, Bridge, Node, initialize, log
    from eagerx.core.graph import Graph
    import eagerx.bridges.openai_gym as eagerx_gym
    import eagerx_examples  # noqa: F401

    # Start roscore & initialize main thread as node
    initialize("eagerx", anonymous=True, log_level=log.INFO)

    # Define object
    pendulum = Object.make("GymObject", "pendulum", env_id="Pendulum-v0", rate=20)

    # Define PID controller & classifier
    pid = Node.make("PidController", "pid", rate=20, gains=[8, 1, 0], y_range=[-4, 4])

    # Define graph (agnostic) & connect nodes
    graph = Graph.create(nodes=[pid], objects=[pendulum])
    graph.connect(source=pendulum.sensors.reward, observation="reward")
    graph.connect(source=pendulum.sensors.done, observation="done")
    graph.connect(source=pendulum.sensors.observation, observation="state")
    # Connect PID
    graph.connect(source=pendulum.sensors.observation, target=pid.inputs.y)
    graph.connect(action="yref", target=pid.inputs.yref)
    graph.connect(source=pid.outputs.u, target=pendulum.actuators.action)

    # Define bridge
    bridge = Bridge.make("GymBridge", rate=20)

    # Initialize Environment (agnostic graph +  bridge)
    env = eagerx_gym.EagerxGym(name=name, rate=20, graph=graph, bridge=bridge)

    # Use stable-baselines
    import stable_baselines3 as sb

    model = sb.SAC("MlpPolicy", env, verbose=1)
    model.learn(total_timesteps=int(eps * eps_length))

    # Evaluate trained policy
    for i in range(eval_eps):
        obs, done = env.reset(), False
        while not done:
            action, _ = model.predict(obs, deterministic=True)
            obs, reward, done, info = env.step(action)
    env.shutdown()


if __name__ == "__main__":
    example_pid_only(name="example", eps=150, eval_eps=20)


eagerx_examples/example_classifier.py
def example_classifier(name, eps, eval_eps, eps_length=200, classifier_epochs=30):
    # EAGERx imports
    from eagerx import Object, Bridge, Node, initialize, log
    from eagerx.core.graph import Graph
    import eagerx.bridges.openai_gym as eagerx_gym
    import eagerx_examples  # noqa: F401

    # Start roscore & initialize main thread as node
    initialize("eagerx", anonymous=True, log_level=log.INFO)

    # Define object
    pendulum = Object.make(
        "GymObject",
        "pendulum",
        sensors=["image", "observation", "reward", "done"],
        env_id="Pendulum-v0",
        rate=20,
        always_render=True,
        render_shape=[28, 28],
    )

    # Define PID controller & classifier
    import os

    dir_path = os.path.dirname(os.path.realpath(__file__))
    classifier = Node.make(
        "Classifier", "classifier", rate=20, cam_rate=20, epochs=classifier_epochs, data=dir_path + "/../data/with_actions.h5"
    )
    pid = Node.make("PidController", "pid", rate=20, gains=[8, 1, 0], y_range=[-4, 4])

    # Define graph (agnostic) & connect nodes
    graph = Graph.create(nodes=[classifier, pid], objects=[pendulum])
    graph.connect(source=pendulum.sensors.reward, observation="reward")
    graph.connect(source=pendulum.sensors.done, observation="done")
    graph.connect(source=classifier.outputs.state, observation="state")
    # Connect Classifier
    graph.connect(source=classifier.outputs.state, target=pid.inputs.y)
    graph.connect(source=pendulum.sensors.image, target=classifier.inputs.image)
    # Connect PID
    graph.connect(action="yref", target=pid.inputs.yref)
    graph.connect(source=pid.outputs.u, target=pendulum.actuators.action)

    # Define bridge
    bridge = Bridge.make("GymBridge", rate=20)

    # Initialize Environment (agnostic graph +  bridge)
    env = eagerx_gym.EagerxGym(name=name, rate=20, graph=graph, bridge=bridge)

    # Use stable-baselines
    import stable_baselines3 as sb

    model = sb.SAC("MlpPolicy", env, verbose=1)
    model.learn(total_timesteps=int(eps * eps_length))

    # Evaluate trained policy
    for i in range(eval_eps):
        obs, done = env.reset(), False
        while not done:
            action, _ = model.predict(obs, deterministic=True)
            action = env.action_space.sample()
            obs, reward, done, info = env.step(action)
    env.shutdown()


if __name__ == "__main__":
    example_classifier(name="example", eps=200, eval_eps=20)


eagerx_examples/example_full_state.py
def example_full_state(name, eps, eval_eps, eps_length=200):
    # EAGERx imports
    from eagerx import Object, Bridge, initialize, log
    from eagerx.core.graph import Graph
    import eagerx.bridges.openai_gym as eagerx_gym
    import eagerx_examples  # noqa: F401

    # Start roscore & initialize main thread as node
    initialize("eagerx", anonymous=True, log_level=log.INFO)

    # Define object
    pendulum = Object.make("GymObject", "pendulum", env_id="Pendulum-v0", rate=20)

    # Define graph (agnostic) & connect nodes
    graph = Graph.create(objects=[pendulum])
    graph.connect(source=pendulum.sensors.observation, observation="observation", window=1)
    graph.connect(source=pendulum.sensors.reward, observation="reward", window=1)
    graph.connect(source=pendulum.sensors.done, observation="done", window=1)
    graph.connect(action="action", target=pendulum.actuators.action, window=1)

    # Define bridge
    bridge = Bridge.make("GymBridge", rate=20)

    # Initialize Environment (agnostic graph +  bridge)
    env = eagerx_gym.EagerxGym(name=name, rate=20, graph=graph, bridge=bridge)

    # Use stable-baselines
    import stable_baselines3 as sb

    model = sb.SAC("MlpPolicy", env, verbose=1)
    model.learn(total_timesteps=int(eps * eps_length))

    # Evaluate trained policy
    for i in range(eval_eps):
        obs, done = env.reset(), False
        while not done:
            action, _ = model.predict(obs, deterministic=True)
            obs, reward, done, info = env.step(action)
    env.shutdown()


if __name__ == "__main__":
    example_full_state(name="example", eps=200, eval_eps=20)


eagerx_examples/classifier.py
from typing import Dict, List, Union, Optional
import matplotlib.pyplot as plt
import numpy as np
import h5py
import warnings
from collections import deque

# IMPORT ROS
import rospy
from std_msgs.msg import Float32MultiArray
from sensor_msgs.msg import Image

# IMPORT EAGERX
import eagerx.core.register as register
from eagerx.utils.utils import Msg
from eagerx.core.entities import Node, Processor, SpaceConverter
from eagerx.core.constants import process


class Classifier(Node):
    @staticmethod
    @register.spec("Classifier", Node)
    def spec(
        spec,
        name: str,
        rate: float,
        cam_rate: float,
        data: str,
        show: bool = False,
        epochs: int = 30,
        window: int = 3,
        state_range: Optional[List[List[float]]] = None,
        process: Optional[int] = process.BRIDGE,
        color: Optional[str] = "grey",
    ):
        """
        Classifier.

        :param spec: Not provided by user.
        :param name: Node name
        :param rate: Rate at which callback is called.
        :param cam_rate: Rate at which images are produced
        :param data: Path to dataste
        :param epochs: Number of epochs to train
        :param show: Flag to plot train results
        :param window: Default window length of image input.
        :param state_range: state range [[min], [max]]. This is not enforced, but merely used for space_converters.
        :param process: {0: NEW_PROCESS, 1: ENVIRONMENT, 2: BRIDGE, 3: EXTERNAL}
        :param color: console color of logged messages. {'black', 'red', 'green', 'yellow', 'blue', 'magenta', 'cyan', 'white', 'grey'}
        :return:
        """
        # Performs all the steps to fill-in the params with registered info about all functions.
        spec.initialize(Classifier)

        # Set default
        state_range = state_range if state_range else [[-1.0, -1.0, -8.0], [1.0, 1.0, 8.0]]

        # Modify default node params
        params = dict(name=name, rate=rate, process=process, color=color, inputs=["image"], outputs=["state"])
        spec.config.update(params)

        # Modify custom node params
        spec.config.update({"cam_rate": cam_rate, "data": data, "show": show, "epochs": epochs})

        # Set window length of images
        spec.inputs.image.window = window

        # Add space_converter to output
        spec.outputs.state.space_converter = SpaceConverter.make(
            "Space_Float32MultiArray", state_range[0], state_range[1], dtype="float32"
        )

    def initialize(self, cam_rate, data, show, epochs):
        self.cam_rate = cam_rate
        # Prepare data
        train, test = prepare_dataset(data, show=show)
        # Train model
        self.classifier = Model()
        self.classifier.train_model(train, test, show=show, epochs=epochs, batch_size=64)

    @register.states()
    def reset(self):
        pass

    @register.inputs(image=Image)
    @register.outputs(state=Float32MultiArray)
    def callback(self, t_n: float, image: Optional[Msg] = None):
        # Prepare image batch
        num_images = len(image.msgs)
        prebatch = bytearray()
        for ros_im in image.msgs:
            if isinstance(ros_im.data, bytes):
                prebatch.extend(ros_im.data)
            else:
                raise NotImplementedError("Only byte arrays are supported.")

        stacked = np.frombuffer(prebatch, dtype=np.uint8).reshape(num_images, 28, 28, -1)

        # Calculate state
        stacked_float32 = change_dtype_np_image(stacked, "float32")
        trig = self.classifier.predict(stacked_float32)
        angles = np.arctan2(
            trig[:, 1],
            trig[:, 0],
        )
        diff = np.diff(angles)
        diff_wrapped = np.arctan2(np.sin(diff), np.cos(diff))
        thdot = diff_wrapped.mean() * self.cam_rate if len(diff_wrapped) > 0 else 0.0
        state = np.array([trig[-1, 0], trig[-1, 1], thdot], dtype="float32")
        return dict(state=Float32MultiArray(data=state))


def prepare_dataset(data_path, show=False, num_examples=12000):
    with h5py.File(data_path, "r") as hf:
        observation = hf["observation.h5"][:]
        state = hf["state.h5"][:]
    print("Loaded observation data: %s" % str(observation.shape))
    print("Loaded state data: %s" % str(state.shape))

    # DATASET PARAMETERS
    observation = observation[:num_examples]
    state = state[:num_examples]

    # DATA PRE-PROCESSING
    # Scale pixel values to a range of 0 to 1 before feeding them to the neural network model.
    observation = observation.astype(np.float32) / 255.0

    # CREATE TEST DATASET
    test_split = 0.2
    if 0 < test_split < 1:
        split_at = int(len(observation) * (1 - test_split))
    else:
        raise ValueError("Must hold-out data as a test dataset. Set parameter 0 < test_split < 1.")
    test_obs = observation[split_at:, :, :, :]
    test_theta = state[split_at:, 0]
    test_trig = np.hstack([np.cos(test_theta)[:, None], np.sin(test_theta)[:, None]])

    # CREATE TRAINING DATASET
    train_obs = observation[:split_at, :, :, :]
    train_theta = state[:split_at, 0]
    train_trig = np.hstack([np.cos(train_theta)[:, None], np.sin(train_theta)[:, None]])

    # VERIFY TRAINING DATA
    # To verify that the data is in the correct format and that you're ready to build and train the network,
    # let's display the first 25 images from the dataset and display the corresponding theta value below each image.
    if show:
        plt.figure(figsize=(10, 10))
        for i in range(25):
            plt.subplot(5, 5, i + 1)
            plt.xticks([])
            plt.yticks([])
            plt.grid(False)
            plt.imshow(train_obs[i])
            plt.xlabel(str(round(train_theta[i] / np.pi, 2)) + "$\\pi$")
    return (train_obs, train_theta, train_trig), (test_obs, test_theta, test_trig)


def change_dtype_np_image(image, dtype):
    if dtype and image.dtype != dtype:
        if image.dtype in ("float32", "float64") and dtype == "uint8":
            image = (image * 255).astype(dtype)
        elif image.dtype == "uint8" and dtype in ("float32", "float64"):
            image = image.astype(dtype) / 255
        else:
            message = "Cannot convert observations from {} to {}."
            raise NotImplementedError(message.format(image.dtype, dtype))
    return image


def show_ros_image(msg):
    import matplotlib.pyplot as plt

    rgb_back = np.frombuffer(msg.data, dtype=np.uint8).reshape(msg.height, msg.width, -1)
    plt.imshow(rgb_back)
    plt.show()


# Torch related
import torch
import torch.nn as nn
from torch.utils.data import TensorDataset, DataLoader


class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = nn.Conv2d(3, 32, kernel_size=(3, 3))
        self.pool = nn.MaxPool2d(2, 2)
        self.fc = nn.Linear(5408, 2)

    def forward(self, x):
        x = self.pool(nn.functional.relu(self.conv(x)))
        x = torch.flatten(x, 1)  # flatten all dimensions except batch
        x = self.fc(x)
        return x

    def predict(self, x):
        if isinstance(x, np.ndarray):
            is_numpy = True
            x = torch.from_numpy(x).permute([0, 3, 1, 2])
        else:
            is_numpy = False
        output = self.forward(x)
        if is_numpy:
            return output.detach().numpy()
        else:
            return output

    def train_model(self, train, test, shuffle=True, batch_size=64, epochs=30, show=False):
        warnings.filterwarnings("ignore")

        train_obs = torch.from_numpy(train[0]).permute([0, 3, 1, 2])
        train_theta = torch.from_numpy(train[1])
        train_trig = torch.from_numpy(train[2])
        test_obs = torch.from_numpy(test[0]).permute([0, 3, 1, 2])
        test_theta = torch.from_numpy(test[1])
        test_trig = torch.from_numpy(test[2])

        # PREPARE TORCH DATASET & DATALOADER
        trainset = TensorDataset(train_obs, train_trig)
        trainloader = DataLoader(trainset, batch_size=batch_size, shuffle=shuffle, num_workers=2)
        testset = TensorDataset(test_obs, test_trig)
        testloader = DataLoader(testset, batch_size=batch_size, shuffle=False, num_workers=2)

        # DEFINE LOSS AND OPTIMIZER
        criterion = nn.MSELoss()
        optimizer = torch.optim.Adam(self.parameters(), lr=0.001)

        for epoch in range(epochs):  # loop over the dataset multiple times
            running_loss = 0.0
            for i, data in enumerate(trainloader, 0):
                # get the inputs; data is a list of [inputs, labels]
                inputs, labels = data

                # zero the parameter gradients
                optimizer.zero_grad()

                # forward + backward + optimize
                outputs = self(inputs)
                loss = criterion(outputs, labels)
                loss.backward()
                optimizer.step()

                # print statistics
                running_loss += loss.item()
                if i % 150 == 149:  # print every 2000 mini-batches
                    rospy.loginfo(f"[epoch={epoch + 1}, mini-batch={i+1}] loss: {running_loss/2000: .5f}")
                    running_loss = 0.0

        print("Finished Training")

        # VALIDATE MODEL
        output = self.predict(test_obs).detach().numpy()
        pred_theta = np.arctan2(output[:, 1], output[:, 0])
        self._validate_model(pred_theta, test_theta, str_model_type="$M^{cnn}$", show=show)

        # MODEL SUMMARY
        if show:
            plt.show()

    def _validate_model(self, pred_theta, test_theta, str_model_type="model_cnn", show=False):
        # EVALUATE MODEL ACCURACY
        # Calculate average error per bin over theta range [-pi, pi]
        test_error = np.abs(test_theta - pred_theta)
        test_error[test_error > np.pi] -= 2 * np.pi
        test_error = np.abs(test_error)
        bins = np.linspace(-np.pi, np.pi, 21)
        digitized = np.digitize(test_theta, bins)
        bin_means = np.array([test_error[digitized == i].mean() for i in range(1, len(bins))])
        if show:
            fig, ax = plt.subplots()
            ax.bar(bins[:-1], bin_means, width=np.diff(bins), edgecolor="black", align="edge")
            ax.set_xlabel("$\\theta$ (rad)")
            ax.set_ylabel("$|\\bar{\\theta} -\\theta|$ (rad)")
            ax.set_title("%s - Average prediction error %s" % (str_model_type, "{:.2e}".format(test_error.mean())))
